File: cashier.cpp
#include "cashier.hpp"
#include <iostream>

Cashier::Cashier()
	: Employee()
	, m_cashRegisterTotal(0.0)
{
	SetPosition("Кассир");
}

Cashier::Cashier(const std::string& fullName, int age, const std::string& contactNumber,
				 const std::string& address, double hourlyRate,
				 const std::string& login, const std::string& password)
	: Employee(fullName, age, contactNumber, address, "Кассир", hourlyRate, login, password)
	, m_cashRegisterTotal(0.0)
{
}

Cashier::Cashier(const Cashier& other)
	: Employee(other)
	, m_cashRegisterTotal(other.m_cashRegisterTotal)
{
}

void Cashier::ProcessPayment(std::shared_ptr<Order> order)
{
	if (order)
	{
		double amount = order->GetTotalAmount();
		m_cashRegisterTotal += amount;

		std::cout << "Кассир " << GetFullName()
				  << " провёл оплату заказа #" << order->GetOrderId()
				  << " на сумму " << amount << " руб."
				  << " (выручка за смену: " << m_cashRegisterTotal << " руб.)" << std::endl;
	}
}

double Cashier::GetCashRegisterTotal() const
{
	return m_cashRegisterTotal;
}

double Cashier::CalculateSalary() const
{
	// Перегрузка без вызова Employee::CalculateSalary().
	// Используем protected‑поля: m_hoursWorked, m_hourlyRate, m_salaryBalance.
	double base = m_hoursWorked * m_hourlyRate;
	double performanceBonus = m_cashRegisterTotal * 0.02; // 2% от проведенной выручки
	double total = base + performanceBonus + m_salaryBalance;

	std::cout << "[Cashier] Расчет зарплаты: базовая = " << base
			  << ", бонус от выручки = " << performanceBonus
			  << ", баланс = " << m_salaryBalance << std::endl;

	return total;
}

File: lab4/Cashier.java
// Кассир, наследуется от Employee
// Отвечает за приём оплаты и суммарную выручку по кассе
public class Cashier extends Employee {
    private double cashRegisterTotal;  // Общая сумма, проведённая через кассу за смену

    // Конструктор по умолчанию
    public Cashier() {
        super();  // Вызов конструктора базового класса
        this.cashRegisterTotal = 0.0;
        setPosition("Кассир");
    }

    // Конструктор с параметрами - демонстрация вызова конструктора базового класса
    public Cashier(String fullName, int age, String contactNumber,
                   String address, double hourlyRate,
                   String login, String password) {
        // Вызов конструктора базового класса Employee с параметрами
        super(fullName, age, contactNumber, address, "Кассир", hourlyRate, login, password);
        this.cashRegisterTotal = 0.0;
    }

    // Проведение оплаты заказа
    public void processPayment(Order order) {
        if (order != null) {
            double amount = order.getTotalAmount();
            cashRegisterTotal += amount;

            System.out.println("Кассир " + getFullName()
                    + " провёл оплату заказа #" + order.getOrderId()
                    + " на сумму " + amount + " руб."
                    + " (выручка за смену: " + cashRegisterTotal + " руб.)");
        }
    }

    // Переопределение метода расчета зарплаты (без вызова базового метода)
    @Override
    public double calculateSalary() {
        // Переопределение без вызова super.calculateSalary()
        // Используем protected-поля напрямую
        double base = hoursWorked * hourlyRate;
        double performanceBonus = cashRegisterTotal * 0.02;  // 2% от проведенной выручки
        double total = base + performanceBonus + salaryBalance;

        System.out.println("[Cashier] Расчет зарплаты: базовая = " + base
                + ", бонус от выручки = " + performanceBonus
                + ", баланс = " + salaryBalance);

        return total;
    }

    // Геттер
    public double getCashRegisterTotal() {
        return cashRegisterTotal;
    }
}

